docker-for-mac: clear the default namespace on start

djs55 opened the pull request:
**- What I did**

Fixed a bug in the `docker-for-mac` blueprint which would cause it to fail to boot if the disk isn't empty i.e. on second and subsequent boots.

**- How I did it**

I delete the directory `/var/lib/containerd/io.containerd.runtime.v1.linux/default` in an `onboot` action before starting `services`. I first tried to modify the `010-containerd` script but unfortunately this is run before the disk is mounted, so the old entries later magically reappear. 

**- How to verify it**

Boot the `docker-for-mac` example twice, without erasing the disk between runs. Without the patch you'll see errors like
```
Starting service: "acpid"
FATA[0000] failed to create task                         error="rpc error: code = Unknown desc = runtime create failed: mkdir /var/lib/containerd/io.containerd.runtime.v1.linux/default/acpid: file exists" service=acpid
```

With the patch, there are no errors and all the services start correctly.

rn commented:
This looks quite hackish to me. If `containerd` is started before things are mounted, shouldn't it be configured to keep state in a temporary filesystem? Or what is the use case for it to keep state on a permanent disk?

justincormack commented:
The use case for it to keep things on persistent storage is if it is being used for Docker, so we dont have to run two containerds. See also #2187

ijc commented:
> I first tried to modify the 010-containerd script but unfortunately this is run before the disk is mounted, so the old entries later magically reappear.

https://github.com/linuxkit/linuxkit/pull/2191 will change that so it could (I think) be done in the `020-containerd` which is where `010-containerd` ends up.

At some point I will add a start-of-day cleanup subcommand to our `service` tool.

But in the meantime this seems fine, LGTM.

justincormack commented:
I am just rebasing #2191 so better to do on top of that in `020-containerd`

ijc commented:
I think I can knock up `service system-init` pretty quickly now I'm done with #2159, so maybe don't hurry to rebase this onto #2191.

ijc commented:
@djs55 does https://github.com/ijc/linuxkit/tree/service-sod-cleanup (which is #2191 + a single commit) solve your issue?

I've only tested by running `service system-init` from a (non-container) `getty` and observing it cleans up everything, but there may be some corner cases or other issues with running from the init script.

djs55 commented:
@ijc probably ;-) here's what I've got now:
```
Welcome to LinuxKit!

NOTE: This system is namespaced.
The namespace you are currently in may not be the root.
login[597]: root login on 'ttyS0'
(ns: getty) linuxkit-025000000001:~# Starting service: "host-timesync-daemon"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/host-timesync-daemon/shim.sock" debug=false module="containerd/tasks" pid=606
Starting service: "ntpd"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/ntpd/shim.sock" debug=false module="containerd/tasks" pid=654
Starting service: "trim-after-delete"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/trim-after-delete/shim.sock" debug=false module="containerd/tasks" pid=702
FATA[0000] failed to create task                         error="rpc error: code = Unknown desc = runtime create failed: OCI runtime create failed: container_linux.go:265: starting container process caused \"process_linux.go:348: container init caused \\\"rootfs_linux.go:57: mounting \\\\\\\"/var/lib/docker\\\\\\\" to rootfs \\\\\\\"/containers/services/trim-after-delete/rootfs\\\\\\\" at \\\\\\\"/var/lib/docker\\\\\\\" caused \\\\\\\"stat /var/lib/docker: no such file or directory\\\\\\\"\\\"\"" service=trim-after-delete
Starting service: "vpnkit-forwarder"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/vpnkit-forwarder/shim.sock" debug=false module="containerd/tasks" pid=731
FATA[0000] failed to create task                         error="rpc error: code = Unknown desc = runtime create failed: OCI runtime create failed: container_linux.go:265: starting container process caused \"process_linux.go:348: container init caused \\\"rootfs_linux.go:57: mounting \\\\\\\"/var/vpnkit\\\\\\\" to rootfs \\\\\\\"/containers/services/vpnkit-forwarder/rootfs\\\\\\\" at \\\\\\\"/port\\\\\\\" caused \\\\\\\"stat /var/vpnkit: no such file or directory\\\\\\\"\\\"\"" service=vpnkit-forwarder
Starting service: "vsudd"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/vsudd/shim.sock" debug=false module="containerd/tasks" pid=759
```
I think several services did start correctly
- `acpid`
- `diagnose`
- `getty`
- `ntpd`
- `vsudd`
but the following failed
- `docker-dfm`
- `vpnkit-forwarder`
- `trim-after-delete`

It looks the same on 1st boot and 2nd. Given this and the fact that some (most) services did start correctly I think the leftover state issue is probably solved. Perhaps the services that failed need some other separate update?

ijc commented:
@djs55 were there any messages about "cleaning up stale services" before the welcome to LinuxKit?

ijc commented:
I put the above into https://github.com/linuxkit/linuxkit/pull/2204

djs55 commented:
@ijc I didn't see a line like that. I added a diff:
```
diff --git a/pkg/containerd/cmd/service/system_init.go b/pkg/containerd/cmd/service/system_init.go
index b3aa8b84..b85e875c 100644
--- a/pkg/containerd/cmd/service/system_init.go
+++ b/pkg/containerd/cmd/service/system_init.go
@@ -82,6 +82,7 @@ func systemInit(args []string) {
                log.WithError(err).Fatal("listing containers")
        }
 
+       log.Infof("Attempting to clean up stale services: %v", ctrs)
        // None of the errors in this loop are fatal since we want to
        // keep trying.
        for _, ctr := range ctrs {
```
and now I see the line
```
Attempting to clean up stale services: []
```
But maybe something else is wrong with my build -- I used to see a bunch of logging related to setting up/ mounting the disk and ethernet device ("waiting for carrier" etc). I don't see any of that now. My console has (omitting most of the initial kernel log):
```
[    3.178784] Freeing unused kernel memory: 156K (ffff9d6a287d9000 - ffff9d6a28800000)
[    3.182712] Freeing unused kernel memory: 1404K (ffff9d6a28aa1000 - ffff9d6a28c00000)
[    3.517858] tsc: Refined TSC clocksource calibration: 3097.768 MHz
[    3.518200] clocksource: tsc: mask: 0xffffffffffffffff max_cycles: 0x2ca70b10e23, max_idle_ns: 440795292750 ns

Starting containerd

Welcome to LinuxKit

                        ##         .
                  ## ## ##        ==
               ## ## ## ## ##    ===
           /"""""""""""""""""\___/ ===
          {                       /  ===-
           \______ O           __/
             \    \         __/
              \____\_______/

INFO[0000] starting containerd boot...                   module=containerd
INFO[0000] loading plugin "io.containerd.content.v1.content"...  module=containerd type=io.containerd.content.v1
INFO[0000] loading plugin "io.containerd.metadata.v1.bolt"...  module=containerd type=io.containerd.metadata.v1
INFO[0000] loading plugin "io.containerd.differ.v1.base-diff"...  module=containerd type=io.containerd.differ.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.containers"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.content"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.diff"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.events"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.healthcheck"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.images"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.namespaces"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.snapshotter.v1.btrfs"...  module=containerd type=io.containerd.snapshotter.v1
WARN[0000] failed to load plugin io.containerd.snapshotter.v1.btrfs  error="expected btrfs, got tmpfs" module=containerd
INFO[0000] loading plugin "io.containerd.snapshotter.v1.overlayfs"...  module=containerd type=io.containerd.snapshotter.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.snapshots"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.monitor.v1.cgroups"...  module=containerd type=io.containerd.monitor.v1
INFO[0000] loading plugin "io.containerd.runtime.v1.linux"...  module=containerd type=io.containerd.runtime.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.tasks"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] loading plugin "io.containerd.grpc.v1.version"...  module=containerd type=io.containerd.grpc.v1
INFO[0000] serving...                                    address="/run/containerd/debug.sock" module="containerd/debug"
INFO[0000] serving...                                    address="/run/containerd/containerd.sock" module="containerd/grpc"
INFO[0000] containerd successfully booted in 0.002387s   module=containerd
Attempting to clean up stale services: []
Starting service: "acpid"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/acpid/shim.sock" debug=false module="containerd/tasks" pid=429
Starting service: "diagnose"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/diagnose/shim.sock" debug=false module="containerd/tasks" pid=475
Starting service: "docker-dfm"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/docker-dfm/shim.sock" debug=false module="containerd/tasks" pid=524
FATA[0000] failed to create task                         error="rpc error: code = Unknown desc = runtime create failed: OCI runtime create failed: container_linux.go:265: starting container process caused \"process_linux.go:348: container init caused \\\"rootfs_linux.go:57: mounting \\\\\\\"/var/vpnkit\\\\\\\" to rootfs \\\\\\\"/containers/services/docker-dfm/rootfs\\\\\\\" at \\\\\\\"/port\\\\\\\" caused \\\\\\\"stat /var/vpnkit: no such file or directory\\\\\\\"\\\"\"" service=docker-dfm
Starting service: "getty"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/getty/shim.sock" debug=false module="containerd/tasks" pid=553

linuxkit-025000000001 login: root (automatic login)

Welcome to LinuxKit!

NOTE: This system is namespaced.
The namespace you are currently in may not be the root.
login[594]: root login on 'ttyS0'
(ns: getty) linuxkit-025000000001:~# Starting service: "host-timesync-daemon"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/host-timesync-daemon/shim.sock" debug=false module="containerd/tasks" pid=603
Starting service: "ntpd"
[    4.527322] clocksource: Switched to clocksource tsc
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/ntpd/shim.sock" debug=false module="containerd/tasks" pid=656
Starting service: "trim-after-delete"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/trim-after-delete/shim.sock" debug=false module="containerd/tasks" pid=699
FATA[0000] failed to create task                         error="rpc error: code = Unknown desc = runtime create failed: OCI runtime create failed: container_linux.go:265: starting container process caused \"process_linux.go:348: container init caused \\\"rootfs_linux.go:57: mounting \\\\\\\"/var/lib/docker\\\\\\\" to rootfs \\\\\\\"/containers/services/trim-after-delete/rootfs\\\\\\\" at \\\\\\\"/var/lib/docker\\\\\\\" caused \\\\\\\"stat /var/lib/docker: no such file or directory\\\\\\\"\\\"\"" service=trim-after-delete
Starting service: "vpnkit-forwarder"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/vpnkit-forwarder/shim.sock" debug=false module="containerd/tasks" pid=729
FATA[0000] failed to create task                         error="rpc error: code = Unknown desc = runtime create failed: OCI runtime create failed: container_linux.go:265: starting container process caused \"process_linux.go:348: container init caused \\\"rootfs_linux.go:57: mounting \\\\\\\"/var/vpnkit\\\\\\\" to rootfs \\\\\\\"/containers/services/vpnkit-forwarder/rootfs\\\\\\\" at \\\\\\\"/port\\\\\\\" caused \\\\\\\"stat /var/vpnkit: no such file or directory\\\\\\\"\\\"\"" service=vpnkit-forwarder
Starting service: "vsudd"
INFO[0000] shim containerd-shim started                  address="/containerd-shim/default/vsudd/shim.sock" debug=false module="containerd/tasks" pid=759
```

ijc commented:
Current code has
```
	for _, ctr := range ctrs {
		log.Infof("Cleaning up stale service: %q", ctr.ID())
```
but I can't remember if I added that and/or pushed it to my branch before I pointed you at it.

I don't see a DHCP container getting started in your logs.

justincormack commented:
replaced by #2204
